### PR TITLE
reporting-server: Created the reporting server package

### LINKED ISSUE #254: Door States Report
## Feature request
 Door States report 

### Description
Include a Door States report in the Report section.

### Implementation considerations
The report should include, the name, doors' mode, DateTime. More fields can be added in case of need.

A table to show the information should be created in the UI. An endpoint and the corresponding filter logic should be implemented in the backend.

### PR DESCRIPTION
**Pull request by matiasbavera**
## What's new

related to #281
fixes #254, #260, #253

Created reporting server. This server will have two functions:
1. Expose an endpoint for FluentD to send log data. This data will be preprocessed and stored in a database (reporting-server-db). We have two POST endpoints, one for storing raw data and other for processing the information coming from `rmf-server`
2. Expose an endpoint where the reporting UI can filter data to build reports.

Created endpoints and reports:
* Raw data
* Dispenser state
* Ingestor state
* Door state
* lift state
* Health status
* Fleet state

Also created and configured the dockerfiles and k8s files for reporting-server and reporting. Configure Fluentd configmap to send data to the reporting-server through Kubernetes services. 

Updated Readme.

Steps to test changes:
* Deploy on Kubernetes FluentD, reporting and reporting server.
* Play with the dashboard for a bit to gather logs
* Go to `example.com/reporting`. 

### How is working now?
image

reportsdeadline

## Self-checks

- [X] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [X] I added unit-tests for new components
- [X] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### TODO 
e2e tests for reporting in general

### CODE REVIEW

**Comment by matiasbavera** [comment]
Logs can be found inside the Fluend pod on  /tmp/all-logs.log/buffer....

A log example:
`
2021-03-31T03:47:16+00:00       kubernetes.var.log.containers.ros2-bridge-66f4678788-6vpbb_default_ros2-bridge-8691aedab8b36d80db6b5860d44412908e7cb0a8b9a11f908f1e5fedd85c5566.log     {"log":"\u001B[32m2021-03-31T03:47:16+00:00 [172.17.0.6] info: received request {\"version\":\"0\",\"method\":\"ros2ServiceCall\",\"params\":{\"service\":{\"type\":\"rmf_building_map_msgs/srv/GetBuildingMap\",\"service\":\"get_building_map\"},\"request\":{}},\"id\":7}\u001B[39m\n","stream":"stdout","docker":{"container_id":"8691aedab8b36d80db6b5860d44412908e7cb0a8b9a11f908f1e5fedd85c5566"},"kubernetes":{"container_name":"ros2-bridge","namespace_name":"default","pod_name":"ros2-bridge-66f4678788-6vpbb","container_image":"rmf-web/ros2-bridge:latest","container_image_id":"docker://sha256:a8aeed9e516b4a807325c249c79d41d93c498b9382a31028cb16d185ce7fb4ee","pod_id":"ab145e4c-b4c7-4a35-9f18-2726d7585251","host":"minikube","labels":{"app":"ros2-bridge","pod-template-hash":"66f4678788"},"master_url":"https://10.96.0.1:443/api","namespace_id":"e192acd4-e6e7-46c2-8514-44a27a367749"}}
`

**Comment by matiasbavera** [comment]
To run minio UI you can run this `kubectl port-forward minio-59ccf98f4d-zcnqb 7000:9000` once you deployed the yaml file. The user is minio and the password is in the minion config file. If you enter here: `http://localhost:7000/`

You'll get this view 
image

**Comment by matiasbavera** [comment]
Hi @koonpeng  I'm having this error: 

```
lib/index.ts(1,28): error TS2307: Cannot find module 'rmf-models' or its corresponding type declarations.
lib/version.ts(2,40): error TS2307: Cannot find module 'rmf-models' or its corresponding type declarations.
```

on the react-component package (GH ci). Is this because it's using an older image? 

On my local machine bootstrap is working correctly.

**Comment by koonpeng** [comment]
> Hi @koonpeng I'm having this error:
> 
> ```
> lib/index.ts(1,28): error TS2307: Cannot find module 'rmf-models' or its corresponding type declarations.
> lib/version.ts(2,40): error TS2307: Cannot find module 'rmf-models' or its corresponding type declarations.
> ```
> 
> on the react-component package (GH ci). Is this because it's using an older image?
> 
> On my local machine bootstrap is working correctly.

Maybe the GH workflows is outdated.

**Comment by matiasbavera** [comment]
> Maybe the GH workflows is outdated.

It's the same as the main branch

File: .venv/.gitignore [hunk 1]
```diff
@@ -1,2 +0,0 @@
-*
-!.gitignore
```
Comment (koonpeng) [inline .venv/.gitignore:2]: Why was this removed, we need the `.venv` dir so that pipenv puts the virtual env locally.

File: .gitignore [hunk 2]
```diff
@@ -4,3 +4,4 @@ npm-debug.log
 
 __pycache__/
 /static
+.idea/
```
Comment (koonpeng) [inline .gitignore:7]: I would recommend to ignore this is your global ignore files, it's not nice if we have to add ignores for every ide.

File: example-deployment/docker/dashboard.dockerfile [hunk 3]
```diff
@@ -20,8 +21,8 @@ COPY --from=0 /root/rmf-web/packages/dashboard/build /usr/share/nginx/html/dashb
 SHELL ["bash", "-c"]
 RUN echo -e 'server {\n\
   location / {\n\
-    root /usr/share/nginx/html;\n\
-    index index.html index.htm;\n\
-    try_files $uri /dashboard/index.html;\n\
+  root /usr/share/nginx/html;\n\
+  index index.html index.htm;\n\
+  try_files $uri /dashboard/index.html;\n\
```
Comment (koonpeng) [inline example-deployment/docker/dashboard.dockerfile:26]: Was this changed by some auto formatter? The indent is wrong.

File: example-deployment/docker/reporting.dockerfile [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# FIXME: we are importing from builder because we are installing packages related to ros on bootstrap.
+FROM rmf-web/builder
```
Comment (koonpeng) [inline example-deployment/docker/reporting.dockerfile:2]: Maybe I missed something, but I don't see any rmf messages used so we shouldn't need to base this off of the base rmf image.

File: example-deployment/docker/reporting.dockerfile [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+FROM nginx:stable
+COPY --from=0 /root/rmf-web/packages/reporting/build /usr/share/nginx/html/reporting
+SHELL ["bash", "-c"]
+RUN echo -e 'server {\n\
+  location / {\n\
+  root /usr/share/nginx/html;\n\
+  index index.html index.htm;\n\
+  try_files $uri /reporting/index.html;\n\
+  }\n\
+  }\n' > /etc/nginx/conf.d/reporting.conf
```
Comment (koonpeng) [inline example-deployment/docker/reporting.dockerfile:27]: indent is wrong

File: example-deployment/k8s/nodeserver.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+      app: node-app
+  template:
+    metadata:
+      labels:
+        app: node-app 
+    spec:
+      containers:
+      - name: node-app
+        imagePullPolicy: Never # remove this in a real deployment
+        image: ekumen/node-web-app:latest
```
Comment (koonpeng) [inline example-deployment/k8s/nodeserver.yaml:34]: Is this a private repo in dockerhub? I can't find it in the public list and it's not built locally.

File: example-deployment/k8s/reporting-server.yaml [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+kind: Service
+metadata:
+  name: reporting-server
+  labels:
+    app: reporting-server
+    tier: app
+spec:
+  type: LoadBalancer
+  selector:
+    app: reporting-server
```
Comment (koonpeng) [inline example-deployment/k8s/reporting-server.yaml:75]: This could be copied from the buggy yaml in the other example k8s files. It needs to have the
`tier: app` selector to prevent conflicts.

File: example-deployment/k8s/reporting-server.yaml [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+kind: Deployment
+metadata:
+  name: reporting-server
+  labels:
+    app: reporting-server
+    tier: app
+spec:
+  selector:
+    matchLabels:
+      app: reporting-server
```
Comment (koonpeng) [inline example-deployment/k8s/reporting-server.yaml:90]: missing `tier` selector

File: example-deployment/k8s/reporting-server.yaml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  template:
+    metadata:
+      labels:
+        app: reporting-server
```
Comment (koonpeng) [inline example-deployment/k8s/reporting-server.yaml:94]: missing `tier` selector

File: example-deployment/k8s/test.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
```
Comment (koonpeng) [inline example-deployment/k8s/test.yaml:1]: leftover?

File: .github/workflows/reporting-server.yml [hunk 11]
```diff
@@ -0,0 +1,41 @@
+name: reporting-server
+on:
+  pull_request:
+    paths:
+      - 'packages/reporting-server/**'
```
Comment (koonpeng) [inline .github/workflows/reporting-server.yml:5]: Is there any integration test between the api server currently? If so we should add the api server to the list of paths.

File: packages/reporting-server/rest_server/default_config.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    "port": 8002,
+    "db_url": "sqlite://:memory:",
+    # "root_path": "/api",
+    # url that rmf-server is being served on.
+    # When being a proxy, this must be the url that rmf-server is mounted on.
+    # E.g. https://example.com/rmf/api/v1
+    "public_url": "http://localhost:8002",
+    # The directory where static files should be stored.
+    "static_directory": "static",
+    "log_level": "DEBUG",  # https://docs.python.org/3.8/library/logging.html#levels
```
Comment (koonpeng) [inline packages/reporting-server/rest_server/default_config.py:13]: Not sure if `DEBUG` should be the default log level.

File: packages/reporting-server/rest_server/default_config.py [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# pylint: disable=line-too-long
+config = {
+    "host": "127.0.0.1",  # ip or hostname to bind the socket to
+    "port": 8002,
+    "db_url": "sqlite://:memory:",
+    # "root_path": "/api",
```
Comment (koonpeng) [inline packages/reporting-server/rest_server/default_config.py:6]: leftover?

File: packages/reporting-server/setup.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python",
+        "Topic :: Software Development",
+    ],
+    install_requires=[
+        "fastapi~=0.63.0",
+        "aiofiles~=0.6.0",
+        "uvicorn[standard]~=0.13.4",
+        "python-socketio~=5.1",
```
Comment (koonpeng) [inline packages/reporting-server/setup.py:23]: does it actually use socketio?

File: packages/reporting-server/setup.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        "rx~=3.1",
```
Comment (koonpeng) [inline packages/reporting-server/setup.py:24]: does it actually use rx?

File: packages/reporting-server/setup.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    keywords=["RMF", "reporting"],
+    classifiers=[
+        "Intended Audience :: Developers",
+        "License :: OSI Approved :: Apache Software License",
+        "Programming Language :: Python",
+        "Topic :: Software Development",
+    ],
+    install_requires=[
+        "fastapi~=0.63.0",
+        "aiofiles~=0.6.0",
```
Comment (koonpeng) [inline packages/reporting-server/setup.py:21]: this is needed only when serving static files, does it actually use it?

**Review by koonpeng — CHANGES_REQUESTED** [review]
Maybe I missed something, but the server doesn't seem like it supports authentication.

File: .github/workflows/reporting-server.yml [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+      - name: tests
+        run: |
+          . /opt/rmf/setup.bash
+          npm run lint
+          npm run test:cov
+          python3 -m pipenv run python -m coverage xml
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v1
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply .github/workflows/reporting-server.yml]: I removed it from here and it's working. Yeah we can create a PR removing all the tokens

File: .gitignore (hunk 2 above)
Comment (matiasbavera) [reply .gitignore]: Yeah, makes sense, thanks!

File: example-deployment/docker/dashboard.dockerfile (hunk 3 above)
Comment (matiasbavera) [reply example-deployment/docker/dashboard.dockerfile]: Yeah, it was prettier. I checkout this file from the main branch

File: .venv/.gitignore (hunk 1 above)
Comment (matiasbavera) [reply .venv/.gitignore]: It was by mistake, sorry. I restored the .venv/ directory with a git checkout from the main branch

File: example-deployment/k8s/nodeserver.yaml (hunk 6 above)
Comment (matiasbavera) [reply example-deployment/k8s/nodeserver.yaml]: Removed the yaml file

File: example-deployment/k8s/test.yaml (hunk 10 above)
Comment (matiasbavera) [reply example-deployment/k8s/test.yaml]: Removing this file. Sorry.

File: .github/workflows/reporting-server.yml (hunk 11 above)
Comment (matiasbavera) [reply .github/workflows/reporting-server.yml:5]: Actually, there is not. The `api-server` doesn't know the existence of the _reporting-server_. It all goes though fluentd

File: packages/reporting-server/rest_server/default_config.py (hunk 12 above)
Comment (matiasbavera) [reply packages/reporting-server/rest_server/default_config.py]: Changed to WARN

File: packages/reporting-server/rest_server/default_config.py (hunk 13 above)
Comment (matiasbavera) [reply packages/reporting-server/rest_server/default_config.py]: Removed. Thanks!

File: packages/reporting-server/setup.py (hunk 14 above)
Comment (matiasbavera) [reply packages/reporting-server/setup.py]: No, leftover. Sorry.

File: packages/reporting-server/setup.py (hunk 16 above)
Comment (matiasbavera) [reply packages/reporting-server/setup.py]: At least no for now, I'll remove it.

File: packages/reporting-server/setup.py (hunk 15 above)
Comment (matiasbavera) [reply packages/reporting-server/setup.py]: No, leftover. Sorry.

**Comment by matiasbavera** [comment]
> Maybe I missed something, but the server doesn't seem like it supports authentication.

You're getting it right, my intention was to add the authentication on a follow-up PR. but I can do that here.

File: example-deployment/docker/reporting.dockerfile (hunk 4 above)
Comment (matiasbavera) [reply example-deployment/docker/reporting.dockerfile:2]: I thought that `react-components` depends on `rmf-models`, which depends of having the models defined (generate-schemas.py). And since `reporting` depends on `react-components` I thought that it was ok using the rmf-web/builder as the base. But I'll try to split it.

File: example-deployment/k8s/reporting-server.yaml (hunk 7 above)
Comment (matiasbavera) [reply example-deployment/k8s/reporting-server.yaml:75]: Oh! Thanks for this!

File: example-deployment/k8s/reporting-server.yaml (hunk 8 above)
Comment (matiasbavera) [reply example-deployment/k8s/reporting-server.yaml:91]: Added!

File: example-deployment/k8s/reporting-server.yaml (hunk 9 above)
Comment (matiasbavera) [reply example-deployment/k8s/reporting-server.yaml:96]: Added thanks!

File: example-deployment/docker/reporting.dockerfile (hunk 4 above)
Comment (koonpeng) [reply example-deployment/docker/reporting.dockerfile:2]: Yeah, but `rmf-models` only needs rmf at the generate phase, right now we are committing the generated files to git to make the build step of other packages simpler, at the expense of making it harder to work on both rmf and rmf-web at the same time (because you need to get your rmf changes merged before you can generate new models for rmf-web). Let me know if you think generating all the models on the fly based on the currently sourced rmf is better.

Comment (matiasbavera) [reply example-deployment/docker/reporting.dockerfile:2]: To use the same version of docker that the other containers use I'll base on `ros:foxy-ros-base-focal`
https://github.com/osrf/docker_images/blob/master/ros/foxy/ubuntu/focal/desktop/Dockerfile. Is that ok?

File: example-deployment/docker/reporting.dockerfile [hunk 18]
```diff
@@ -1,9 +1,16 @@
 # FIXME: we are importing from builder because we are installing packages related to ros on bootstrap.
-FROM rmf-web/builder
+FROM ros:foxy-ros-base-focal
```
Comment (koonpeng) [inline example-deployment/docker/reporting.dockerfile:2]: does the FIXME still apply?

I missed that this is a multi stage build in the previous review, so it doesn't really matter if we use the rmf builder or other images. Both the previous version and this should be fine.

**Review by koonpeng — APPROVED** [review]
